src/Simplex/Chat/Mobile/Shared.hs

```diff
@@ -1,19 +1,26 @@
 module Simplex.Chat.Mobile.Shared where
 
-import qualified Data.ByteString as B
-import Data.ByteString.Internal (ByteString (PS), memcpy)
+import Data.ByteString.Internal (ByteString (..), memcpy)
 import Foreign.C (CInt, CString)
-import Foreign (Ptr, Word8, newForeignPtr_, plusPtr)
-import Foreign.ForeignPtr.Unsafe
+import Foreign
 
 type CJSONString = CString
 
+type JSONByteString = ByteString
+
 getByteString :: Ptr Word8 -> CInt -> IO ByteString
 getByteString ptr len = do
   fp <- newForeignPtr_ ptr
-  pure $ PS fp 0 $ fromIntegral len
+  pure $ BS fp $ fromIntegral len
 
 putByteString :: Ptr Word8 -> ByteString -> IO ()
-putByteString ptr bs@(PS fp offset _)  = do
-  let p = unsafeForeignPtrToPtr fp `plusPtr` offset
-  memcpy ptr p $ B.length bs
+putByteString ptr (BS fp len) =
+  withForeignPtr fp $ \p -> memcpy ptr p len
+
+newCStringFromBS :: ByteString -> IO CString
```

Comment:
Why not [unsafePackCString](https://hackage.haskell.org/package/bytestring-0.12.0.2/docs/Data-ByteString-Unsafe.html#v:unsafePackCString) ?

Reply:
not sure why they do it like this everywhere? You mean to avoid any copying?

Reply:
but ByteString can be garbage collected before CString is used.

Reply:
And the rest from Data.ByteString.Unsafe...

Reply:
I see, the code gives out `Ptr`s for outside code.